<!doctype html>
<html lang="en">

	<head>
		<meta charset="utf-8">

		<title>Python intro - Jan 15 2105</title>

		<meta name="description" content="An introduction to Python for Coding">
		<meta name="author" content="Wayne ecatur">

		<meta name="apple-mobile-web-app-capable" content="yes" />
		<meta name="apple-mobile-web-app-status-bar-style" content="black-translucent" />

		<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">

		<link rel="stylesheet" href="css/reveal.css">
		<link rel="stylesheet" href="css/theme/black.css" id="theme">

		<!-- Code syntax highlighting -->
		<link rel="stylesheet" href="lib/css/zenburn.css">

		<!-- Printing and PDF exports -->
		<script>
			var link = document.createElement( 'link' );
			link.rel = 'stylesheet';
			link.type = 'text/css';
			link.href = window.location.search.match( /print-pdf/gi ) ? 'css/print/pdf.css' : 'css/print/paper.css';
			document.getElementsByTagName( 'head' )[0].appendChild( link );
		</script>

		<!--[if lt IE 9]>
		<script src="lib/js/html5shiv.js"></script>
		<![endif]-->
	</head>

	<body>

		<div class="reveal">

			<!-- Any section element inside of this container is displayed as a slide -->
			<div class="slides">
				<section>
					<h1>Introduction to Coding in Python</h1>
					<h3>January 15th, 2015</h3>
					<p>
						<small>Created by <a href="http://proteopedia.org/wiki/index.php/User:Wayne_Decatur">Wayne Decatur</a> for Feng Lab Group meeting</small>
					</p>
				</section>

				<section data-markdown>
					<script type="text/template">
						## Workshop Infrastructure
						[the hub is here](http://bit.ly/FengPyCode)

						Easy to arrive at two ways:

						- write down `http://bit.ly/FengPyCode`
						- search for `Wayne Decatur proteopedia`, follow github and look for repo for Feng lab workshop

					</script>
				</section>


				<section data-markdown>
					<script type="text/template">
						### Today is not really about Python

						- working productively
						- automation
						- reusable code
						- modular design
						- develop smart, test often
						- programming concepts and algorithmic thinking
						- troubleshooting
						- collaboration
						- reproducibility --> reproducible science
						- open source and open science

							... Datapocalypse (yours and the field in general)

					</script>
				</section>

				<section data-markdown>
					<script type="text/template">
						### Prep reading?

						comments?

						highly recommend [Practical Computing for Biologists by Haddock and Dunn ](http://practicalcomputing.org/)

						_I am a nomad_

					</script>
				</section>









				<section>
					<h2>Why code in Python?</h2>
						<p class="fragment">Maybe a better question at this point is <q>&ldquo;why code?&rdquo;</q></p>
				</section>





				<section data-markdown>
					<script type="text/template">
						## Why code?

						- quickly perform repetitive tasks accurately
						- automate pipelines and workflows
						- allows implementing algorithms and leads to algorithmic thinking that empowers you in problem-solving
						- can help in creating text and visual assets

						Adds up to better data handling and greater productivity.

					</script>
				</section>





				<section>
					<img width = "480px" height = "650px" data-src="HHMIcover.png" alt="winter 2013 HHMI magazine">
				</section>

				<section>
					<img width = "960px" height = "650px" data-src="HHMIinside_sidebyside.png" alt="inside cover pages">
				</section>

				<section data-transition="fade">
					<img width = "480px" height = "650px" data-src="HHMIinsidecover.png" alt="my image">
				</section>

				<section data-transition="fade">
					<img width = "480px" height = "650px" data-src="HHMIwcode.png" alt="my image with code example">
				</section>

				<section data-markdown>
					<script type="text/template">
						## When to code

						Don't forget regular expressions and basic shell commands can handle a lot of your text and file manipulation needs.

When you reach the limit of what you can easily do with those, then you need to code.

>For example, recently wanted something that produced mRNA sequences from a list of NCBI's mRNA entries in FASTA form, such as [Schizosaccharomyces pombe 972h- ribonuclease MRP complex subunit, mRNA](http://www.ncbi.nlm.nih.gov/nuccore/NM_001019799.2).


					</script>
				</section>

				<section>
					<h2>How to approach that?</h2>
						<p class="fragment">Easy `find and replace` won't work because changes all `T's` and may change useful information.</p>
						<p class="fragment">Regular expression may have worked if I figured out how to do backward assertion?</p>
				</section>

				<section data-markdown data-transition="fade">
					<script type="text/template">
						### So we are getting closer...

						Look closer at the example input [here](https://gist.githubusercontent.com/fomightez/ba6762fdb38dd6ff4377/raw/0de93d0a561a0037d36f46237fa2d9655408cabb/fasta_example)



					</script>
				</section>



				<section data-transition="slide" data-background="#b5533c" data-background-transition="zoom">
					<h2>Pseudocode for the problem</h2>

					<pre><code data-trim contenteditable>
Go through all the lines input and decide what to do with them.

Want to skip lines beginning with >.

For all the other lines, we want replace 't' with 'u' elsewhere.
					</code></pre>

				</section>



				<section data-markdown>
					<script type="text/template">
						## Why code in Python?

						Of course there's lots of options for languages

						Examples:

						- Perl
						- Java
						- [Antha](http://www.antha-lang.org/)
						- many others
					</script>
				</section>




				<section data-transition="slide" data-background="#4d7e65" data-background-transition="zoom">
					<img data-src="http://imgs.xkcd.com/comics/python.png" alt="Python can make you fly">
				</section>



				<section data-markdown>
					<script type="text/template">
						### Python

						- robust, versatile, fully mature language, beyond basic 'scripts', despite it being a 'scripting language' since interpreted and not compiled. Lots of room to grow and learn. can even do object oriented programming.

						- [very high-level language](https://darynholmes.wordpress.com/2008/04/07/very-high-level-language-%E2%80%93-so-what/)

						- Widely available and portable

						- Lots of support and resources

						- Highly suited to text processing. See [here](https://www.ibm.com/developerworks/library/l-python5/), [here](http://pythonforbiologists.com/index.php/introduction-to-python-for-biologists/python-for-biologists-introduction/) and [here](https://bcrc.bio.umass.edu/courses/spring2012/micbio/micbio660/content/python-lab-6-manipulating-files-and-processing-text) for example) which makes up a lot of genomics/ngs work



					</script>
				</section>

				<section data-markdown>
					<script type="text/template">
						PYTHON _continued_

						- Popular among those using data of many, many kinds. Especially scientists!

						- Current 'popular kid' among bioinformaticians (and biologists in general) for a general language

						- You are a fan of British humor and yearn to be known as a Phythonista.

					</script>
				</section>




				<section data-transition="slide" data-background="#b5533c" data-background-transition="zoom">
					<h2>Python is a Very High Level Language</h2>
					<pre><code data-trim contenteditable>
import random


numbers = range (1,50)
chosen = []

while len(chosen) < 6:
	number = random.choice(numbers)
	numbers.remove(number)
	chosen.append(number)

chosen.sort()

print "This week's numbers are", chosen
print "The bonus ball is", random.choice(numbers)
					</code></pre>
					<p>Code example borrowed from <a href="https://darynholmes.wordpress.com/2008/04/07/very-high-level-language-%E2%80%93-so-what/">here</a>.</p>
				</section>

				<section data-markdown>
					<script type="text/template">
						###Python versatility

						- Django - Instagram, Mozilla site, Pinterest, [Schmitt Lab site](http://schmittlab.pythonanywhere.com/)
						- Pymol for structural biology

					</script>
				</section>




				<section data-markdown>
					<script type="text/template">
						#### Python 2.x or 3.x?
I vote for 2.x.

- Typical for many scientific modules and packages to find 2.x supported and 3.x coming along. (See [here](https://jakevdp.github.io/blog/2013/01/03/will-scientists-ever-move-to-python-3/) for an overview, albeit outdated. Situation is vastly improved now.)


					</script>
				</section>

								<section data-markdown>
					<script type="text/template">
						#### Python 2.x or 3.x?

Examples:

From [Python for next-generation sequencing](http://www.digitalbiologist.com/2013/06/python-next-gen-sequencing.html)
>The Python library links in this article are for Python 2.7 which is the version currently supported by most of the major scientific and numerical Python packages, as well as being the most commonly used release in the scientific Python community.

Another at [Sourcelair](https://www.sourcelair.com/guides/start/python#python-versions-supported)

					</script>
				</section>

												<section data-markdown>
					<script type="text/template">
						#### Python 2.x or 3.x?

Not a big deal.

Biggest difference for most people is print command.

See [key differences between Python 2.7.x and Python 3.x](http://nbviewer.ipython.org/github/rasbt/python_reference/blob/master/tutorials/key_differences_between_python_2_and_3.ipynb)

For basic things it is minor.

Don't let it be a hurdle to getting started.
					</script>
				</section>


				<section data-markdown>
					<script type="text/template">
						## Where to code and run Python?

						see the docs


					</script>
				</section>

				<section>
					<h4>Sidebar: Running Python for today</h4>
						<p><small>Reminder that approach used here today was just to save on installation issue. See <a href="http://ivory.idyll.org/blog/2013-swc-arizona.html">here</a> for an example of how it can hinder getting started.</small></p>

						<p><small>For exploring and initial development cloud-approach is fine. Suggest though as you move towards real development of useful programs, you code in a text editor or IDE on your machine and then run in the cloud or even on your machine, depending on how nomadic you are or computing power/recources needed.</small></p>

						<p><small>Why do I say do the code development locally?
- main reason is reliability. You can easily undo and are more likely to have active backup mechanisms available on your computer or through linking with Dropbox etc., just like you would when you do work on a Microsoft Word document or presentation. Later you'd be smart to integrate git/github into your code development workflow, but that serves a different purpose (organizing development and collaboration) other than direct recovery of your most recent edits.</small></p>
<li><small>sites go down for spells</small></li>
<li><small>companies fold</small></li>
<li><small>internet not always accessible</small></li>


<p><small>Bottom line: get started, then work out your coding and running style/workflow. Won't be same for all things obviously. Examples from a nomad- mostly PythonAnywhere; large jobs in the cloud unless no money than can use lab desktop (at the cost of convenience).</small></p>
				</section>


				<section data-markdown>
					<script type="text/template">
						###Getting started

						Today we'll use Sourcelair and Sagemath Cloud

[SourceLair](https://www.sourcelair.com/home)

[Guide to Python on Sourcelair](https://www.sourcelair.com/guides/start/python#introduction)

[Sagemath Cloud](https://cloud.sagemath.com)

					</script>
				</section>




				<section data-markdown>
					<script type="text/template">
						###Using Sourcelair to run Python Interactively

						Wading in...

						_We'll break it up with some real world examples_
					</script>
				</section>

				<section data-markdown>
					<script type="text/template">
						###Break I: Running Other's Python code
					</script>
				</section>


				<section data-markdown>
					<script type="text/template">
						###Running Python in an IPython Notebook

						- [Sagemath Cloud](https://cloud.sagemath.com)
					</script>
				</section>


				<section data-markdown>
					<script type="text/template">
						###Modules/Packages/libraries

						- [Matplotlib](http://matplotlib.org/gallery.html)
						- [Biopython](http://biopython.org/DIST/docs/tutorial/Tutorial.html)
						- RE
						- urllib2
					</script>
				</section>


				<section data-markdown>
					<script type="text/template">
						###APIs can now be your friend

						- [Plotly](https://plot.ly/~wayne461/42/file-size-distribution-of-all-105222-protein-data-bank-entries-as-of-jan-7-2015/)
						- [NCBI Entrez server](http://www.ncbi.nlm.nih.gov/books/NBK25501/) via Biopython
						- [Intermine](http://yeastmine.yeastgenome.org/yeastmine/begin.do) ([YeastMine](http://yeastmine.yeastgenome.org/yeastmine/begin.do))
						- [Ensembl Rest API](http://www.ncbi.nlm.nih.gov/pubmed/25236461)

					</script>
				</section>

				<section data-markdown>
					<script type="text/template">
						###Break II: Use some APIs
					</script>
				</section>




				<section>
					<h2>Putting it all together</h2>

					<p>Starting with <a href="https://gist.githubusercontent.com/fomightez/964a77bf8d5f89e443a3/raw/6b419814cc3a6c5a25ef79e827fb339af6f0ee43/latlon3.py">this example</a> , can you write a program to convert DNA fasta to mRNA?</p>
						<p class="fragment">example, <a href="https://gist.github.com/fomightez/f73300b7be3a982d6b09">DNA_to_RNAsimple.py</a></p>
				</section>







				<section data-markdown>
					<script type="text/template">
						###Tips for getting started

						- Start small

						- Test often

						- Document

						- Develop modularly as much as you can.






					</script>
				</section>

				<section data-markdown>
					<script type="text/template">
						####Tips for getting started
						_...continued_

						-Leave time for developing . Consider time to write the code vs time to do it by hand. Factor in the high degree of accuracy you gain, over doing it by hand, when you code.


					</script>
				</section>

				<section data-markdown>
					<script type="text/template">
						####Tips for getting started
						_...continued_

						Advice from [here](http://nbviewer.ipython.org/github/swcarpentry/2013-04-az/blob/master/notebooks/10-introducing-bird-counting-RICH.ipynb)

>Some basic principles of this kind of computing
Try to make your code easy to read. Rather than commenting everything to death, make your function and variable names descriptive.
Test out your functions on small data sets where you know the answer.
The third time you find yourself performing the same query or merge, write a function to do it instead. Corrollary: don't overplan in advance! Only do this the third time!
					</script>
				</section>





				<section data-markdown>
					<script type="text/template">
						###Places for questions
- Try Google, probably will lead you to one of my listed resources or...
- Stackoverflow for general scripting and computing
- [Biostars](https://www.biostars.org)

Can take effort and some educating of yourself to get to the point you can make a meaningful query with the right terminology to find anything related.

When you know enough to know what is covered elsewhere and what isn't, then ask uncovered questions at those sites.
					</script>
				</section>







				<section data-markdown>
					<script type="text/template">
						####Suggested immediate next steps

[Practical Computing for Biologists by Haddock and Dunn ](http://practicalcomputing.org/) - overall great resource for all aspects of computing

Install Enthought's Canopy or Anaconda Python distribution on you system. They have free academic use versions.
Supplement or expand on that with utilization of PythonAnywhere and/or Domino Data Labs.

Run other people's Python code to make you more productive.


Just use python and code! - Try things. Run others code. Look at others code (Github and blogs) and using resources on the internet (in particular Biostars and Stackoverflow) can get you far.


					</script>
				</section>


								<section data-markdown>
					<script type="text/template">
						####Suggested immediate next steps
						_...continued_

I suggest as you work considering everything you do and see where repetition is dragging you down or where ability to be more exhausative would be a benefit.

Another workshop?

Try a course eventuallly?

Watch some time soon:

[A History of Bioinformatics (in the Year 2039)](https://www.youtube.com/watch?v=uwsjwMO-TEA)
					</script>
				</section>
















				<section style="text-align: left;">
					<h1>THE END</h1>
					<p>

					</p>
				</section>







			</div>

		</div>

		<script src="lib/js/head.min.js"></script>
		<script src="js/reveal.js"></script>

		<script>

			// Full list of configuration options available at:
			// https://github.com/hakimel/reveal.js#configuration
			Reveal.initialize({
				controls: true,
				progress: true,
				history: true,
				center: true,

				transition: 'slide', // none/fade/slide/convex/concave/zoom

				// Optional reveal.js plugins
				dependencies: [
					{ src: 'lib/js/classList.js', condition: function() { return !document.body.classList; } },
					{ src: 'plugin/markdown/marked.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
					{ src: 'plugin/markdown/markdown.js', condition: function() { return !!document.querySelector( '[data-markdown]' ); } },
					{ src: 'plugin/highlight/highlight.js', async: true, condition: function() { return !!document.querySelector( 'pre code' ); }, callback: function() { hljs.initHighlightingOnLoad(); } },
					{ src: 'plugin/zoom-js/zoom.js', async: true },
					{ src: 'plugin/notes/notes.js', async: true }
				]
			});

		</script>

	</body>
</html>
